### ISSUE TITLE
Serializing a `tf.keras.Model` in half-precision

### ISSUE THREAD

**Issue by sayakpaul** [keras-team/keras#17381]
**System information**.
- Have I written custom code (as opposed to using a stock example script provided in Keras): Yes
- OS Platform and Distribution (e.g., Linux Ubuntu 16.04): Colab
- TensorFlow installed from (source or binary): Binary
- TensorFlow version (use command below): 2.11.0
- Python version: 3.8.16
- Bazel version (if compiling from source):
- GPU model and memory:
- Exact command to reproduce:

**Describe the problem**.

After setting the precision policy to be `mixed_float16` the serialized model is not in half-precision. 

Is this expected? For larger models (Stable Diffusion, for example), allowing users to serialize models in half-precision can be quite beneficial. 

**Describe the current behavior**.

The model gets serialized in full precision.

**Describe the expected behavior**.

Users should have a way to serialize a model in half-precision. 

**[Contributing](https://github.com/keras-team/keras/blob/master/CONTRIBUTING.md)**.

- Do you want to contribute a PR? (yes/no): no
- If yes, please read [this page](https://github.com/keras-team/keras/blob/master/CONTRIBUTING.md) for instructions
- Briefly describe your candidate solution(if contributing):

**Standalone code to reproduce the issue**.

Colab Gist: https://colab.research.google.com/gist/sayakpaul/c7abb6f8057ba21857985de5e26a22d8/scratchpad.ipynb

**Comment by sushreebarsa** [comment]
@SuryanarayanaY I was able to replicate the issue, please find the [gist](https://colab.research.google.com/gist/sushreebarsa/796992c3fc0dd2ec23d1da246d865152/scratchpad.ipynb) here.
Thank you!

**Comment by SuryanarayanaY** [comment]
Hi @sayakpaul ,
Apologies for delayed response.Setting the policy to 'mixed_float16' or 'mixed_bfloat16', makes the` compute dtype` to be `float16` or `bfloat16` and the `variable dtype to be float32`.
In the attached code of yours doesn't checked the serialized model dtypes. I tested the same code and after `model.save()`, I have reloaded the model and checked the dtype which is same as initial model.

For initial model:
```
model Compute dtype: float16
model Variable dtype: float32
```
For serialized model after reloading:
```
reloaded_model Compute dtype: float16
reloaded_model Variable dtype: float32
```
Please refer to attached [gist](https://colab.research.google.com/gist/SuryanarayanaY/1f0a9484420dd0ab1fa29f89558365e4/keras-17381-r1.ipynb).

This is what you looking right? Please confirm if anything else missing here.

Thankyou!

**Comment by sayakpaul** [comment]
I want to serialize the model weights in FP16.

**Comment by SuryanarayanaY** [comment]
@sayakpaul ,
Policy will be applicable to computations only. As Model weights are Variables they will remain as `float32` dtype only which is default.Please refer to documentation [link1](https://www.tensorflow.org/api_docs/python/tf/keras/mixed_precision/Policy) and [link2](https://www.tensorflow.org/guide/mixed_precision).The main purpose of Policy is to improve performance by making computations in float16 when set global_policy to `mixed_float16 ` or `mixed_bfloat16`.

**Comment by SuryanarayanaY** [comment]
@sayakpaul ,
This is not a serialization issue.Model variables(weights) always will be either `float32` (by default) or `float64` if set global_policy to `float64`. Model weights will not be converted to half precision since it affects Accuracy of Model.

The minimum precision for model `variables` is `float32` by default.Even if you set '`mixed_float16`' for global policy and layer dtype also, the model `variables(weights)` will remains as `float32` only.You can only change `model.variables` to `float64` by setting the policy to `float64`.
Please refer to attached demo gist
You can check the `model.variable_dtype` for your model before and after serialization also it will remain same.Please report if you find any descripencies.

**Comment by sayakpaul** [comment]
This is a bummer that even if I wanted to have the model weights in FP16, I cannot do it.

**Comment by SuryanarayanaY** [comment]
@sayakpaul ,
I have found one API which can

**Comment by SuryanarayanaY** [comment]
@sayakpaul ,
Sorry there has been some correction regarding model.weights default dtype.Actually I have digged more and found an [API](https://www.tensorflow.org/api_docs/python/tf/keras/backend/set_floatx) `tf.keras.backend.set_floatx` which actually controls the default dtype of Keras. As we know the model variable dtypes are 'float32' by default we can set this to 'float16' using `tf.keras.backend.set_floatx('float16')`.In this way we can minimize the precision level in computations wherever required to improve performance.But even with this API if we use dtype='float16' for all layers but after building model the Variables still be stored as 'float32' only.Please refer to attached gist [here](https://colab.research.google.com/gist/SuryanarayanaY/ff868cc2c0b4e2c4d287c0e1fa7901e3/mixed_precision_policy_demo-r1.ipynb).But we can however use weights as 'float16' which can improve the model performance by reducing computation time and resources at cost of Accuracy if required.

**Comment by sayakpaul** [comment]
@SuryanarayanaY thanks!

Could you please include a section comparing the model sizes in your Colab so that size benefits are evident?

**Comment by innat** [comment]
@SuryanarayanaY 
> Policy will be applicable to computations only. As Model weights are Variables they will remain as float32 dtype only which is default.

That's understandable.  
But as it might be useful in certain cases (larger model), what if a new arugment is introduced in the saving API, for example: `keras.callbacks.ModelCheckpoint(saving_policy='float16', ..)`/ `model.save(saving_policy='float16`)`, etc! cc @rchao  @reedwm
